.github/workflows/check-changelog.yml

```diff
@@ -0,0 +1,52 @@
+name: Check Changelog
+# This check makes sure that the changelog is properly updated
+# when a PR introduces a change in a test file.
+# To bypass this check, label the PR with "No Changelog Needed".
+on:
+  pull_request:
+
+jobs:
+  check:
+    runs-on: ubuntu-latest
+    if: ${{ contains(github.event.pull_request.labels.*.name, 'No Changelog Needed') == 0 }}
+    steps:
+      - name: Get PR number and milestone
+        run: |
+          echo "PR_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+          echo "MILESTONE_TITLE=${{ github.event.pull_request.milestone.title }} >> $GITHUB_ENV
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: '0'
+      - name: Check the changelog
+        run: |
+          set -xe
+          changed_files=$(git diff --name-only origin/master)
+          # Changelog should be updated only if tests have been modified
+          if [[ "$changed_files" =~ tests ]]
+          then
+            all_changelogs=$(cat ./doc/whats_new/v*.rst)
+            if [[ "$all_changelogs" =~ :pr:\`$PR_NUMBER\` ]]
+            then
+              echo "Changelog has been updated."
+              # If the pull request is milestoned check the correspondent changelog
+              if exist -f ./doc/whats_new/v${MILESTONE_TITLE:0:4}.rst
+              then
+                expected_changelog=$(cat ./doc/whats_new/v${MILESTONE_TITLE:0:4}.rst)
+                if [[ "$expected_changelog" =~ :pr:\`$PR_NUMBER\` ]]
+                then
+                  echo "Changelog and milestone correspond."
+                else
+                  echo "Changelog and milestone do not correspond."
+                  echo "If you see this error and there is already a changelog entry then make sure that"
+                  echo "the tagged milestone for the PR (if any) and the changelog name properly match."
```

Comment:
IIUC, in this branch:

- there is already a changelog entry
- there is a tag for the milestone, otherwise the file wouldn't exist

So we can remove "and tehre is already a cahngelog entry" and "(if any)"